Sprite scale: Preserve performance

einarf opened the pull request:
The `Sprite.scale` changes made the property 4-5 times slower. Instead we introduce methods to handle uniform scaling.

einarf committed:
Initial work

einarf committed:
More fixes + tests

einarf committed:
Example fixes

DragonMoffon commented:
Id like to try one last thing before we go ahead. Since the biggest cost with the Vec2 is around creation and GC I wonder if making a slotted murable Vec-like with support for piecewise operations would be performant enough. Like working with lists it may have a few hickups, but Ill implement `__copy__`

DigiDuncan commented:
> Id like to try one last thing before we go ahead. [...]

I'm curious to see this before we push this one through -- the DX is a tad clunky and it won't be forward-compatible with 3.1. I'm fine with it if Dragon's proposal doesn't get done/isn't up to snuff, though, I agree it's a better option then lying to the user as to what the underlying scale is.

einarf commented:
The amount of things that broke due to this change was minuscule so I'm not too worried. The new methods are also a lot faster than the scale property can be. Especially when adding and multiplying scale simply because it's one callable less compared to `+=` and `*=`. 

This also needs to be checked before merge. Ensure all subclasses of `BasicSprite` is typed to allow a tuple as well.
https://github.com/pythonarcade/arcade/issues/2399

DragonMoffon committed:
Updated scale method names, and reimplimented scalar setting of scale

The names we had for the scale methods didn't feel right for arcade/python so they have been changed to be more readable.

While the concern about confusing the user is valid I think the change is worth it. During the interim before we do a larger re-write this is a nicer developer experience than using a method. Most developers set the sprite scale once, and then don't change it in the future.

the `isinstance` check for `AsFloat` is no-where near the cost of creating a vector.

In my ideal world sprite would not have a scale property at all, and instead scale is a method that changes the size of the sprite.

DragonMoffon committed:
Fix examples and tutorials

DragonMoffon committed:
fixing typing error

DragonMoffon committed:
linting and type pass

DragonMoffon commented on arcade/examples/sprite_explosion_particles.py:
Still has wrong method will fix soon

DragonMoffon committed:
last place the old method was hiding

DigiDuncan reviewed (approved):
As much as the functions feel a tad clunky still, the speed can't be denied, and I'm more concerned about the return of `scale = float` for DX reasons. Seems good to me until 3.1+.

DragonMoffon committed:
perf tests and improved scale setter

DragonMoffon committed:
comparing func vs unpacking and random numbers in perf test

DragonMoffon committed:
arcade scale incorect type unit test

DragonMoffon committed:
Merge branch 'development' into scale-tweaks
